lua/caching_handlers.lua

```diff
@@ -15,22 +15,50 @@ function caching_handlers._extract_ids_from_uri(uri, pattern)
     return ids
 end
 
--- Callback to save response to cache, to be executed after the response has been sent
-function caching_handlers._post_request_callback(response, request_info, cacheability_info)
+-- Function to compute the id field used in the cache.
+function caching_handlers._get_cache_key(request_info, cacheability_info)
     local ids = {'null'}
     if cacheability_info.cache_entry.enable_id_extraction then
-        ids = caching_handlers._extract_ids_from_uri(
-            request_info.normalized_uri,
-            cacheability_info.cache_entry.pattern
-        )
+        if request_info.request_method ~= 'POST' then
+            ids = caching_handlers._extract_ids_from_uri(
+                request_info.normalized_uri,
+                cacheability_info.cache_entry.pattern
+            )
+        elseif request_info.request_body ~= nil then
+            ids = {
+                    spectre_common.get_id_from_req_body(
+                        cacheability_info.cache_entry.post_body_id,
+                        request_info.request_body
+                    )
+            }
+        else
+            error ("POST endpoint with id_extraction enabled has no request body" .. request_info.normalized_uri)
+        end
     end
+    return ids
+end
+
+-- Function to get uri (cache_key) in the cache.
+-- For post request with normalized body it returns uri..body
+function caching_handlers._get_cache_uri(request_info)
+    local cache_uri = request_info.normalized_uri
+    if request_info.request_method == 'POST' and request_info.normalized_body ~= nil then
```

Comment:
could we do this check only once (maybe when we verify if a request is cacheable or not) and not repeat it everywhere?

Reply:
i've moved the POST method check to only `determine_if_cacheable` , but had to do the check in `normalize_body` method so I don't add the body (if present) for a GET method.